src/java/org/apache/cassandra/index/sai/disk/v3/BlockTermsWriter.java

```diff
@@ -0,0 +1,723 @@
+
+    private FileValidation.Meta createCRCMeta(long rootFP, IndexComponent comp) throws IOException
+    {
+        try (IndexInput input = indexDescriptor.openPerIndexInput(comp, context, segmented))
+        {
+            return FileValidation.createValidate(rootFP, input);
+        }
+    }
+
+    public long pointCount()
+    {
+        return pointId;
+    }
+
+    public int blockCount()
+    {
+        return postingsBlockCount;
+    }
+
+    public boolean add(ByteComparable term, long rowid) throws IOException
+    {
+        copyBytes(term, spare);
+        return add(spare.get(), rowid, null);
+    }
+
+    public void add(ByteComparable term, long rowid, TermRowIDCallback callback) throws IOException
+    {
+        copyBytes(term, spare);
+        add(spare.get(), rowid, callback);
+    }
+
+    /**
+     * Block of prefix bytes cache.
+     */
+    private static class BytesBlockCache
+    {
+        private final BytesRefArray bytesArray = new BytesRefArray(Counter.newCounter(false));
+        private final BytesRefBuilder spare = new BytesRefBuilder();
+        private boolean sameTerms = true;
+
+        public int count()
+        {
+            return bytesArray.size();
+        }
+
+        public BytesRef minTerm()
+        {
+            return bytesArray.get(spare, 0);
+        }
+
+        private void add(BytesRef term)
+        {
+            if (bytesArray.size() > 0)
+            {
+                final BytesRef prevTerm = bytesArray.get(spare, bytesArray.size() - 1);
+                if (!prevTerm.equals(term))
+                    sameTerms = false;
+            }
+            bytesArray.append(term);
+        }
+
+        private void reset()
+        {
+            sameTerms = true;
+            bytesArray.clear();
+        }
+    }
+
+    /**
+     * Write the bytes block to disk.
+     */
+    private void writeBytesBlock(final BytesBlockCache bytesCache) throws IOException
+    {
+        writeBytesBlockPrevTerm.clear();
+        for (int x = 0; x < bytesCache.count(); x++)
+        {
+            final BytesRef term = bytesCache.bytesArray.get(bytesCache.spare, x);
+            if ((x & TERMS_DICT_BLOCK_MASK) == 0)
+            {
+                lastBytesBlockFP = termsOut.getFilePointer();
+                termBlockFPs.add(lastBytesBlockFP);
+
+                termsOut.writeVInt(term.length);
+                termsOut.writeBytes(term.bytes, term.offset, term.length);
+            }
+            else
+            {
+                final int prefixLength = BlockTerms.bytesDifference(writeBytesBlockPrevTerm.get(), term);
+                final int suffixLength = term.length - prefixLength;
+
+                termsOut.writeByte((byte) (Math.min(prefixLength, 15) | (Math.min(15, suffixLength - 1) << 4)));
+
+                if (prefixLength >= 15)
+                    termsOut.writeVInt(prefixLength - 15);
+                if (suffixLength >= 16)
+                    termsOut.writeVInt(suffixLength - 16);
```

Comment:
magic numbers, would be good to document where they're coming from

Reply:
This is from `SortedTermsWriter` which is partially copied from Lucene, which you reviewed prior.  

Here's the Lucene code in `Lucene80DocValuesConsumer`:

```
 } else {
    final int prefixLength = StringHelper.bytesDifference(previous.get(), term);
    final int suffixLength = term.length - prefixLength;
    assert suffixLength > 0; // terms are unique

    data.writeByte((byte) (Math.min(prefixLength, 15) | (Math.min(15, suffixLength - 1) << 4)));
    if (prefixLength >= 15) {
      data.writeVInt(prefixLength - 15);
    }
    if (suffixLength >= 16) {
      data.writeVInt(suffixLength - 16);
    }
    data.writeBytes(term.bytes, term.offset + prefixLength, term.length - prefixLength);
  }
```

Reply:
Lucene has no comments so I'll need to figure out what's going on.